pkg/controller/ephemeraljob/ephemeraljob_controller.go

```diff
@@ -0,0 +1,353 @@
+	var requeueMsg string
+	defer func() {
+		if err != nil {
+			klog.Warningf("Failed to process EphemeralJob %v err %v, elapsedTime %v", request.Name, time.Since(start), err)
+		} else if res.RequeueAfter > 0 {
+			klog.Infof("Finish to process EphemeralJob %v, elapsedTime %v, RetryAfter %v: %v", request.Name, time.Since(start), res.RequeueAfter, requeueMsg)
+		} else {
+			klog.Infof("Finish to process EphemeralJob %v, elapsedTime %v", request.Name, time.Since(start))
+		}
+	}()
+
+	job := &appsv1alpha1.EphemeralJob{}
+	err = r.Get(context, request.NamespacedName, job)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		klog.Errorf("Failed to get ephemeral job %s,", job.Name)
+		return reconcile.Result{}, err
+	}
+
+	// The Job has been finished
+	if job.Spec.TTLSecondsAfterCreated != nil && job.CreationTimestamp.Add(time.Duration(*job.Spec.TTLSecondsAfterCreated)*time.Second).Before(time.Now()) {
+		klog.Infof("Deleting EphemeralJob %s/%s for ttlSecondsAfterFinished", job.Namespace, job.Name)
+		if err = r.Delete(context, job); err != nil {
+			return reconcile.Result{}, fmt.Errorf("Delete ephemeral job error: %v. ", err)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	if skipReconcile(job) {
+		return reconcile.Result{}, nil
+	}
+
+	targetPods, matches, err := r.filterPods(job) // matches 表示选择匹配的数量, targetPods表示要操纵的列表
+	if err != nil {
+		klog.Errorf("Failed to get ephemeral job %s/%s related target pods: %v", job.Namespace, job.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	// 计算calculateStatus
+	if err := r.calculateStatus(job, targetPods, matches); err != nil {
+		klog.Errorf("Error calculate ephemeral job %s/%s status: %v", job.Namespace, job.Name, err)
+		return reconcile.Result{}, err
+	}
+	klog.Infof("Sync calculate job %s status: %v", job.Name, job.Status)
+
+	if job.Status.Phase == appsv1alpha1.EphemeralJobPause {
+		return reconcile.Result{Requeue: true, RequeueAfter: 0}, r.updateJobStatus(job)
+	}
+
+	if err := r.syncTargetPods(job, targetPods); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, r.updateJobStatus(job)
+}
+
+func (r *ReconcileEphemeralJob) filterPods(job *appsv1alpha1.EphemeralJob) ([]*v1.Pod, int, error) {
+	selector, err := util.GetFastLabelSelector(job.Spec.Selector)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	opts := &client.ListOptions{
+		Namespace:     job.Namespace,
+		LabelSelector: selector,
+	}
+
+	podList := &v1.PodList{}
+	if err := r.List(context.TODO(), podList, opts); err != nil {
+		return nil, 0, err
+	}
+
+	// Ignore inactive pods
+	var targetPods []*v1.Pod
+	var matches int
+	for i := range podList.Items {
+		if kubecontroller.IsPodActive(&podList.Items[i]) {
```

Comment:
what if pod list sequence changes across multiple reconciles? for example if there is scaling out activities, we may have newly created pods which get matched and being injected ephemeral containers , which may violate the replicas restriction. we may require sort the pod first, and if there is informer cache update latency, we may need to introduce expectations